CubicChunksAPI/src/main/java/io/github/opencubicchunks/cubicchunks/api/world/storage/ICubicStorage.java

```diff
@@ -0,0 +1,248 @@
+     */
+    @Nonnull
+    default NBTBatch readBatch(PosBatch positions) throws IOException {
+        //default implementation: issue reads individually, but in parallel
+        try {
+            return new NBTBatch(
+                    positions.columns.parallelStream().collect(Collectors.toConcurrentMap(pos -> pos, pos -> {
+                        try {
+                            return this.readColumn(pos);
+                        } catch (IOException e) {
+                            throw new UncheckedIOException(e);
+                        }
+                    })),
+                    positions.cubes.parallelStream().collect(Collectors.toConcurrentMap(pos -> pos, pos -> {
+                        try {
+                            return this.readCube(pos);
+                        } catch (IOException e) {
+                            throw new UncheckedIOException(e);
+                        }
+                    })));
+        } catch (UncheckedIOException e) {
+            throw e.getCause(); //rethrow original exception
+        }
+    }
+
+    /**
+     * Writes the NBT data to the column at the given position.
+     *
+     * @param pos the column's position
+     * @param nbt the column's NBT data
+     */
+    void writeColumn(ChunkPos pos, NBTTagCompound nbt) throws IOException;
+
+    /**
+     * Writes the NBT data to the cube at the given position.
+     *
+     * @param pos the cube's position
+     * @param nbt the cube's NBT data
+     */
+    void writeCube(CubePos pos, NBTTagCompound nbt) throws IOException;
+
+    /**
+     * Writes the NBT data for multiple cubes+columns at once.
+     *
+     * @param batch a {@link NBTBatch} containing the cube+column positions and the NBT data to write to each
+     */
+    default void writeBatch(NBTBatch batch) throws IOException {
+        //default implementation: issue writes individually, but in parallel
+        try {
+            batch.columns.entrySet().parallelStream().forEach(entry -> {
+                try {
+                    this.writeColumn(entry.getKey(), entry.getValue());
+                } catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            });
+            batch.cubes.entrySet().parallelStream().forEach(entry -> {
+                try {
+                    this.writeCube(entry.getKey(), entry.getValue());
+                } catch (IOException e) {
+                    throw new UncheckedIOException(e);
+                }
+            });
+        } catch (UncheckedIOException e) {
+            throw e.getCause(); //rethrow original exception
+        }
+    }
+
+    /**
+     * Iterates over all the columns that exist in the world.
+     *
+     * @param callback the callback function to run
+     */
+    void forEachColumn(Consumer<ChunkPos> callback) throws IOException;
```

Comment:
I don't know for sure whether current implementation of RegionLib provides a thread safe way to do this

Reply:
actually i'm not sure either, i just assumed it was thread-safe because everything else in regionlib is. i'll dig around in the code and see how it works

Reply:
looks like it should be fine. it opens all the regions on a separate thread, and only reads the headers to see if the cube/column exists.